executor/join.go

```diff
@@ -406,18 +626,122 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
 		innerRows = append(innerRows, innerRow)
 	}
 
-	resultBuffer.rows, resultBuffer.err = e.resultGenerator.emit(outerRow, innerRows, resultBuffer.rows)
+	resultBuffer.rows, resultBuffer.err = e.resultGenerators[0].emit(outerRow, innerRows, resultBuffer.rows)
 	if resultBuffer.err != nil {
 		resultBuffer.err = errors.Trace(resultBuffer.err)
 		return false
 	}
 	return true
 }
 
+func (e *HashJoinExec) joinMatchedOuterRow2Chunk(workerID int, outerRow chunk.Row, chk *chunk.Chunk) error {
+	buffer := e.hashJoinBuffers[workerID]
+	hasNull, joinKey, err := e.getJoinKeyFromChkRow(e.outerKeys, outerRow, buffer.bytes)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if hasNull {
+		err = e.resultGenerators[workerID].emitToChunk(outerRow, nil, chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	}
+	innerPtrs := e.hashTable.Get(joinKey)
+	if len(innerPtrs) == 0 {
+		err = e.resultGenerators[workerID].emitToChunk(outerRow, nil, chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	}
+	innerRows := make([]chunk.Row, 0, len(innerPtrs))
+	for _, b := range innerPtrs {
+		ptr := *(*chunk.RowPtr)(unsafe.Pointer(&b[0]))
+		matchedInner := e.innerResult.GetRow(ptr)
+		innerRows = append(innerRows, matchedInner)
+	}
+
+	err = e.resultGenerators[workerID].emitToChunk(outerRow, chunk.NewSliceIterator(innerRows), chk)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (e *HashJoinExec) buildNewJoinResult(workerID int, joinResult *execWorkerResult) (bool, *execWorkerResult) {
+	joinResult = &execWorkerResult{
+		cycleChan: e.joinChkResourceCh[workerID],
+	}
+	ok := true
+	select {
+	case <-e.closeCh:
+		return false, joinResult
+	case joinResult.chk, ok = <-e.joinChkResourceCh[workerID]:
+		if !ok {
+			return false, joinResult
+		}
+	}
+	return true, joinResult
+}
+
+func (e *HashJoinExec) join2Chunk(workerID int, outerChk *chunk.Chunk, joinResultChkBuffer *chunk.Chunk, joinResult *execWorkerResult, selected []bool) (ok bool, _ *execWorkerResult) {
+	var err error
+	selected, err = expression.VectorizedFilter(e.ctx, e.outerFilter, outerChk, selected)
+	if err != nil {
+		joinResult.err = errors.Trace(err)
+		return false, joinResult
+	}
+	for i, matched := range selected {
```

Comment:
no need to bring in `matched`, it increases the difficulty to understand the code behavior. 
```go
for i := range selected
```